### PR TITLE
feat: improve Tooltip pseudo-system by allowing other mods to access it

### PR DESCRIPTION
**Pull request by TaroEld**
Adding tooltips for new UI elements is tiresome. This adds a simple system where you pass a table with your new tooltips to a function, and you can then use the tooltips in the UI by passing a string corresponding to the table structure.

### CODE REVIEW

File: msu/ui/tooltip/classes/abstract_tooltip.nut [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		this.Text = _text;
+	}
+
+	function getUIData()
+	{
+		local ret = [
+			{
+				id = 1,
+				type = "title",
+				text = typeof this.Title == "string" ? this.Title : this.Title()
```
Comment (Enduriel) [inline msu/ui/tooltip/classes/abstract_tooltip.nut:19]: as above

File: msu/ui/tooltip/classes/abstract_tooltip.nut [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+			},
+			{
+				id = 2,
+				type = "description",
+				text = typeof this.Text == "string" ? this.Text : this.Text()
```
Comment (Enduriel) [inline msu/ui/tooltip/classes/abstract_tooltip.nut:24]: as above

File: msu/ui/tooltip/classes/abstract_tooltip.nut [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+{
+	Title = null;
+	Text = null;
+
+	constructor(_title, _text)
+	{
+		::MSU.requireOneFromTypes(["string", "function"], _title, _text);
+		this.Title = _title;
+		this.Text = _text;
+	}
```
Comment (Enduriel) [inline msu/ui/tooltip/classes/abstract_tooltip.nut:11]: what about converting title and text to a lambda function if it's a string during initialization so you don't have to worry about this?
```squirrel
this.Title = typeof _title == "string" ? @() _title : _title;
this.Text = typeof _text == "string" ? @() _text : _text;
```
That way you can always assume it's a function. 'Parse, don't validate' xD

File: msu/ui/tooltip/classes/abstract_tooltip.nut [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+				id = 1,
+				type = "title",
+				text = typeof this.Title == "string" ? this.Title : this.Title()
+			},
+			{
+				id = 2,
+				type = "description",
+				text = typeof this.Text == "string" ? this.Text : this.Text()
+			},
+		]
```
Comment (Enduriel) [inline msu/ui/tooltip/classes/abstract_tooltip.nut:26]: missing ;

**Review by Enduriel — CHANGES_REQUESTED** [review]
So Midas' complaining does work!

File: msu/ui/tooltip/classes/abstract_tooltip.nut (hunk 3 above)
Comment (TaroEld) [reply msu/ui/tooltip/classes/abstract_tooltip.nut]: Has been implemented with getters and setters.

File: msu/ui/tooltip/classes/abstract_tooltip.nut (hunk 4 above)
Comment (TaroEld) [reply msu/ui/tooltip/classes/abstract_tooltip.nut]: yes

File: msu/systems/tooltips/abstract_tooltip.nut [hunk 5]
```diff
@@ -0,0 +1,20 @@
+::MSU.Class.Tooltip <- class
+{
+	Data = null;
+
+	constructor(_data = null)
```
Comment (LordMidas) [inline msu/systems/tooltips/abstract_tooltip.nut:5]: space before and after args

File: msu/systems/tooltips/abstract_tooltip.nut [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	{
+		this.setData(_data);
+	}
+
+	function setData(_data)
```
Comment (LordMidas) [inline msu/systems/tooltips/abstract_tooltip.nut:10]: space before and after args

File: msu/systems/tooltips/abstract_tooltip.nut [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	{
+		::MSU.requireOneFromTypes(["table", "null"], _data);
+		this.Data = _data == null ? {} : _data;
+	}
+
+	function getUIData(_data)
```
Comment (LordMidas) [inline msu/systems/tooltips/abstract_tooltip.nut:16]: space before and after args

File: msu/systems/tooltips/abstract_tooltip.nut [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+	Data = null;
+
+	constructor(_data = null)
+	{
+		this.setData(_data);
+	}
+
+	function setData(_data)
+	{
+		::MSU.requireOneFromTypes(["table", "null"], _data);
```
Comment (LordMidas) [inline msu/systems/tooltips/abstract_tooltip.nut:12]: Better way to write this imo:
```squirrel
if (_data == null) this.Data = {};
else
{
     ::MSU.requireTable(_data);
     this.Data = _data;
}
```

File: msu/systems/tooltips/tooltips/basic_tooltip.nut [hunk 9]
```diff
@@ -0,0 +1,51 @@
+::MSU.Class.BasicTooltip <- class extends ::MSU.Class.Tooltip
+{
+	Title = null;
+	Text = null;
+
+	constructor(_title, _text, _data = null)
```
Comment (LordMidas) [inline msu/systems/tooltips/tooltips/basic_tooltip.nut:6]: space before and after args. I will stop posting this comment further. Please go through the entire PR and fix all instances of this.

File: msu/systems/tooltips/tooltips/basic_tooltip.nut [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	function getTitle(_data)
+	{
+		return this.Title.getinfos().parameters.len() == 1 ? this.Title() : this.Title(_data);
+	}
+
+	function getText(_data)
+	{
+		return this.Text.getinfos().parameters.len() == 1 ? this.Text() : this.Text(_data);
```
Comment (LordMidas) [inline msu/systems/tooltips/tooltips/basic_tooltip.nut:32]: as above

File: msu/systems/tooltips/tooltips/basic_tooltip.nut [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+	function setText(_text)
+	{
+		::MSU.requireOneFromTypes(["string", "function"], _text);
+		this.Text = typeof _text == "string" ? @() _text : _text;
+	}
+
+	function getTitle(_data)
+	{
+		return this.Title.getinfos().parameters.len() == 1 ? this.Title() : this.Title(_data);
```
Comment (LordMidas) [inline msu/systems/tooltips/tooltips/basic_tooltip.nut:27]: Instead of this, why don't you simply force that when someone passes a function using `setTitle` that the function must take a single argument i.e. `_data`? This is cleaner imo as compared to doing this `getinfos()` check.

File: msu/systems/tooltips/tooltips/basic_tooltip.nut [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	function getText(_data)
+	{
+		return this.Text.getinfos().parameters.len() == 1 ? this.Text() : this.Text(_data);
+	}
+
+	function getUIData(_data)
+	{
+		local ret = [
```
Comment (LordMidas) [inline msu/systems/tooltips/tooltips/basic_tooltip.nut:37]: No need to create a local variable. Replace `local` with `return` and remove the return line at the bottom.

File: msu/systems/tooltips/tooltips_system.nut [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+				_currentTable[key] <- value;
+			}
+		}
+	}
+
+	function getTooltip( _modID, _identifier )
+	{
+		local fullKey = split(_identifier, ".");
+		local currentTable = this.Mods[_modID];
+		while (fullKey.len() > 0)
```
Comment (LordMidas) [inline msu/systems/tooltips/tooltips_system.nut:43]: `remove` is rather expensive afaik. A more efficient implementation is as follows:
```squirrel
for (local i = 0; i < fullKey.len(); ++i)
{
	currentTable = currentTable[fullKey[i]];
}
```

File: scripts/config/msu/classes/weighted_container.nut [hunk 14]
```diff
@@ -218,6 +218,46 @@
 		return null;
 	}
 
+	function rollTopX(_top, _exclude = null )
```
Comment (LordMidas) [inline scripts/config/msu/classes/weighted_container.nut:221]: Move this stuff to a separate PR.

File: msu/hooks/ui/screens/tooltip/tooltip_events.nut [hunk 15]
```diff
@@ -39,16 +39,22 @@
...
 	}
 
+	o.onQueryMSUTooltipData <- function( _data )
+	{
+		return ::MSU.System.Tooltips.getTooltip(_data.modId, _data.elementId).getUIData(_data);
+	}
+
 	local general_queryUIElementTooltipData = o.general_queryUIElementTooltipData;
-	o.general_queryUIElementTooltipData = function( _entityId, _elementId, _elementOwner )
+	o.general_queryUIElementTooltipData = function( _entityId, _elementId, _elementOwner ) // mod settings should be converted to use onQueryMSUTooltipData
```
Comment (LordMidas) [inline msu/hooks/ui/screens/tooltip/tooltip_events.nut:48]: Please fix as necessary because I don't understand what is going on here. @Enduriel

**Review by LordMidas — CHANGES_REQUESTED** [review]

File: msu/systems/tooltips/tooltips/basic_tooltip.nut (hunk 11 above)
Comment (Enduriel) [reply msu/systems/tooltips/tooltips/basic_tooltip.nut:27]: That was how it originally was, this is nicer for the user imo

File: msu/hooks/ui/screens/tooltip/tooltip_events.nut (hunk 15 above)
Comment (Enduriel) [reply msu/hooks/ui/screens/tooltip/tooltip_events.nut]: changes to this have been reverted, and the hook has been marked as deprecated

**Review by Enduriel — APPROVED** [review]

**Comment by LordMidas** [comment]
Closing without merge as a squashed version of this branch has been merged into development manually.
